jhbuild/modtypes/npm.py

```diff
@@ -0,0 +1,155 @@
+        """
+        buildscript.set_action(_('Installing Package Dependencies'), self)
+
+        source = os.path.join(self.get_srcdir(buildscript), 'package.json')
+        dest = os.path.join(self.get_builddir(buildscript), 'package.json')
+        if not os.path.exists(self.get_builddir(buildscript)):
+            os.makedirs(self.get_builddir(buildscript))
+        shutil.copyfile(source, dest)
+        self.npm(buildscript, 'install', npmargs='--prefix %s' % self.get_builddir(buildscript))
+
+    def do_build(self, buildscript):
+        """ run bundler to build and pack
+        """
+        self.prepare_installroot(buildscript)
+        args = {
+            'prefix': self.get_destdir(buildscript) + buildscript.config.prefix
+        }
+        self.bundlerargs = self.bundlerargs % args
+        getattr(self, self.bundler, self.default_bundler)(buildscript)
+
+    do_build.depends = [PHASE_INSTALL_DEPENDENCIES]
+    do_build.error_phases = [PHASE_FORCE_CHECKOUT]
+
+    def do_install(self, buildscript):
+        """ install the output files to installdir
+        """
+        buildscript.set_action(_('Installing'), self)
+        self.process_install(buildscript, self.get_revision())
+
+    do_install.depends = [PHASE_BUILD]
+    do_install.error_phases = [PHASE_FORCE_CHECKOUT]
+
+    def get_srcdir(self, buildscript):
+        """ source file folder
+        """
+        return self.branch.srcdir
+
+    def get_builddir(self, buildscript):
+        return os.path.join(buildscript.config.buildroot, self.name)
+
+    def get_npmcmd(self, config):
+        if self.npmcmd:
+            return self.npmcmd
+        return 'npm'
+
+    def get_npmargs(self, buildscript):
+        npmargs = ' %s %s' % (self.npmargs, self.config.module_npmargs.get(self.name, self.config.npmargs))
+        return self.eval_args(npmargs).strip()
+
+    def npm(self, buildscript, target='', npmargs=None, env=None):
+        npmcmd = os.environ.get('NPM', self.get_npmcmd(buildscript.config)) # TODO
+
+        if npmargs is None:
+            npmargs = self.get_npmargs(buildscript)
+
+        extra_env = (self.extraenv or {}).copy()
+        for k in (env or {}):
+            extra_env[k] = env[k]
+
+        cmd = '{npm} {npmargs} {target}'.format(npm=npmcmd,
+                                                npmargs=npmargs,
+                                                target=target)
+        buildscript.execute(cmd, cwd=self.get_builddir(buildscript), extra_env=extra_env)
+
+    # bundlers               
+    def default_bundler(self, buildscript):
+        raise CommandError('unknown bundler')
+
+    def webpack(self, buildscript):
+        webpack = os.path.join(self.get_builddir(buildscript), 'node_modules/.bin', 'webpack')
+        cmd = '%s %s' % (webpack, self.bundlerargs)
+        buildscript.execute(cmd, cwd=self.get_srcdir(buildscript))
+
+    def xml_tag_and_attrs(self):
+        return 'npm', [('id', 'name', None),]
```

Comment:
You will need `npmargs`, `bundler`, `bundlerargs` here. This is for dumping the jhbuild modules file